key/aziot-key-openssl-engine-shared/build/engine.c

```diff
@@ -18,8 +18,14 @@ int aziot_key_openssl_engine_shared_bind(ENGINE* e, const char* id);
  *
  * Unfortunately this trick means this crate cannot be compiled for tests since the linker will see duplicate symbols, so every invocation of
  * `cargo test --all` or `cargo clippy --tests --all` has to also exclude this crate with `--exclude`.
+ *
+ * For openssl 3.0 the `__asm__(".symver")` approach causes `ld` to complain about duplicate symbols.
```

Comment:
If the symbols aren't exported, the engine won't be usable. openssl 1.1 vs 3.0 won't make a difference here, because the macros are defined exactly the same way, ie it's the same symbols that need to be exported.

You can double-check with `nm -D engine.so` if the symbols are exported or not.

Are you sure you were seeing the duplicate symbols warning with a `make test-release` build and not something like `cargo test` ? If it was the latter then the comment above this explains why the duplicate symbols happen.

Reply:
I was seeing the duplicate symbol warning when trying to do `make PACKAGE_RELEASE=1 PACKAGE_VERSION=1.4.1 V=1 deb`.  I'll double-check.

Reply:
Hmm, strange.

Yeah, if you are sure you get the error with `make deb`, then check if the engine.so with the `__asm__` lines removed still exports the `bind_engine` and `v_check` symbols. If it really does export those symboles then it would work, but it'd be a mystery why that happens with 3.0 but not with 1.1 given the macros seem to be identical, and the fact that the symbols were not getting exported was a problem with rustc and not with openssl.

Reply:
Hmm... I'm not seeing any symbols if I do `nm -D libaziot_key_openssl_engine_shared.so | grep bind_engine`.

But, I'm also getting duplicate symbol errors if I leave the engine.c file untouched.  Here's the command I ran:

```bash
docker run -it --rm -v "$(realpath ~/src/iot-identity-service):/src" \
    -e 'ARCH=amd64' \
    -e 'PACKAGE_VERSION=1.4.0' \
    -e 'PACKAGE_RELEASE=1' \
    ubuntu:22.04 '/src/ci/package.sh'
```

And the resulting error:

```
...
error: linking with `cc` failed: exit status: 1
  |
  = note: "cc" "-Wl,--version-script=/tmp/rustcQ8VKfQ/list" "-m64" "/tmp/rustcQ8VKfQ/symbols.o" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/aziot_key_openssl_engine_shared.aziot_key_openssl_engine_shared.dfc62e90-cgu.0.rcgu.o" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/aziot_key_openssl_engine_shared.aziot_key_openssl_engine_shared.dfc62e90-cgu.1.rcgu.o" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/aziot_key_openssl_engine_shared.aziot_key_openssl_engine_shared.dfc62e90-cgu.10.rcgu.o" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/aziot_key_openssl_engine_shared.aziot_key_openssl_engine_shared.dfc62e90-cgu.11.rcgu.o" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/aziot_key_openssl_engine_shared.aziot_key_openssl_engine_shared.dfc62e90-cgu.12.rcgu.o" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/aziot_key_openssl_engine_shared.aziot_key_openssl_engine_shared.dfc62e90-cgu.13.rcgu.o" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/aziot_key_openssl_engine_shared.aziot_key_openssl_engine_shared.dfc62e90-cgu.14.rcgu.o" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/aziot_key_openssl_engine_shared.aziot_key_openssl_engine_shared.dfc62e90-cgu.15.rcgu.o" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/aziot_key_openssl_engine_shared.aziot_key_openssl_engine_shared.dfc62e90-cgu.2.rcgu.o" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/aziot_key_openssl_engine_shared.aziot_key_openssl_engine_shared.dfc62e90-cgu.3.rcgu.o" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/aziot_key_openssl_engine_shared.aziot_key_openssl_engine_shared.dfc62e90-cgu.4.rcgu.o" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/aziot_key_openssl_engine_shared.aziot_key_openssl_engine_shared.dfc62e90-cgu.5.rcgu.o" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/aziot_key_openssl_engine_shared.aziot_key_openssl_engine_shared.dfc62e90-cgu.6.rcgu.o" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/aziot_key_openssl_engine_shared.aziot_key_openssl_engine_shared.dfc62e90-cgu.7.rcgu.o" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/aziot_key_openssl_engine_shared.aziot_key_openssl_engine_shared.dfc62e90-cgu.8.rcgu.o" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/aziot_key_openssl_engine_shared.aziot_key_openssl_engine_shared.dfc62e90-cgu.9.rcgu.o" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/aziot_key_openssl_engine_shared.4uytss4ehckkywsf.rcgu.o" "-Wl,--as-needed" "-L" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps" "-L" "/tmp/aziot-identity-service-1.4.0/target/release/deps" "-L" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/build/aziot-key-openssl-engine-shared-e763449c4ea9bf98/out" "-L" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/build/aziot-key-openssl-engine-950e6f910dc97734/out" "-L" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/build/openssl-sys2-1ecd412231953aa1/out" "-L" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,-Bstatic" "-Wl,--whole-archive" "-laziot_key_openssl_shared_engine_wrapper" "-Wl,--no-whole-archive" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libaziot_key_openssl_engine-3fc2df14767b1cd3.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libaziot_key_client-e151afd6f6366504.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libaziot_key_common_http-386aafcfd9410b71.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libhttp_common-aa536d2159557dad.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libserde_json-30cd81ff86c038cd.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libryu-b1a64d482ee8fb27.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libhyper_openssl-5d916e031f8c72ea.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/liblinked_hash_set-6ed567ab718724c9.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/liblinked_hash_map-9e3554fee5db2970.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libtower_layer-4cd95fa30ed94670.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libparking_lot-217e0b0864bf2ca9.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libparking_lot_core-8359453dc2998b96.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libsmallvec-383d02b8e6149f1c.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/liblock_api-2fd6c57d43876436.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libscopeguard-20056e42e927ca2b.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libhyper_proxy-297633320df5dc9d.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libheaders-54cc91a96eea6041.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libsha1-62cc54a4c8455d28.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libcpufeatures-6cb79141dfe06232.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libdigest-af1e7da3413bf904.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libblock_buffer-5c01c07976ea3294.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libcrypto_common-799653fe47d55bac.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libgeneric_array-35c7cbdf52fc30b4.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libtypenum-f4c0806ba70e8a1d.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libmime-693424db13d06c5f.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libheaders_core-82910076943bcc11.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libbase64-f5ef83bad373743c.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libfutures-cccf19f4b95708ac.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libfutures_executor-b7dbffdbb8d4eadf.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libtokio_openssl-ebef9e9cba0fca73.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/liburl-e7912308b310ec61.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libidna-2b0314ea5ceceb72.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libunicode_normalization-135d99d328c8922e.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libtinyvec-641fa7cb5c47e03b.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libtinyvec_macros-82fbdfd202c42795.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libunicode_bidi-61b88d82b72a92d0.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libform_urlencoded-3ac79fc58f7ba920.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libpercent_encoding-fc7db89b7498772f.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libmatches-163c174974d8490e.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libhyper-04a999d92b0fa6a7.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libwant-733a0996c1391d60.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libtry_lock-871f8a32ecd4eeea.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libhttparse-4ded89f32f844d3b.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libh2-cc4892dd88bdeedc.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libindexmap-cb546a6fa47bcd7d.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libhashbrown-de58b707eb419386.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libtokio_util-16e5da966a120af7.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libtower_service-e9f1cd97893caa3d.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libtracing-58a1486f92a0055d.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libtracing_core-fe51506eb3e1fb80.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libhttp_body-4c8bc2365028a936.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libhttpdate-a06ba372acf48bf7.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libhttp-5383b8ca54b6d46e.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libitoa-ad569e520b4022a6.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libfnv-c6824ab0c291d1f7.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libtokio-ab50ca834b79ef72.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libnum_cpus-1fd78e2d523fbe95.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libsocket2-a084386caf935374.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libbytes-8a4893ac85cd4650.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libmio-ccecc49f3d8c967a.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libnix-f3a65432d747da28.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libmemoffset-ebf708fffb62bb1b.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libfutures_util-e7445cef628f53d1.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libmemchr-a68a60acdb3eebd7.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libfutures_io-9a71ea596d0eee11.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libslab-ef3d7ac3b26386a0.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libfutures_channel-5fcea9ada85b2c61.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libpin_project_lite-39b54a6c9490936b.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libfutures_sink-ea98f0f3d10efe1f.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libfutures_task-116043b16e6ee7f4.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libpin_utils-5df07a926ff31a44.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libfutures_core-d44fd726aae7ed91.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libaziot_key_common-14877c019bc37385.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libserde-c35deeae1f8612a0.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libopenssl2-4dece0fc7cc236a7.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libopenssl_sys2-d808bccdf70921d9.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libopenssl-77097ce093fc849f.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libonce_cell-b0e8e973b7663233.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libbitflags-3b6ecce377a6aaad.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libforeign_types-9fd8dc78e45fceb7.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libforeign_types_shared-4a9d445533aa0863.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/liblog-6172e032849674bd.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libcfg_if-092951c3dbdf50b2.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libopenssl_errors-787e9d1b6661eeb1.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libopenssl_sys-ff039b1d2e627490.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/liblibc-70cb204fd8252664.rlib" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libcfg_if-97b2ec5240e8f1dd.rlib" "-Wl,--start-group" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-8f1929c73c3f8167.rlib" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_abort-2ef255b89354cca0.rlib" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libobject-b886fd10c5a7c7c0.rlib" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libmemchr-9d7c322d48daa475.rlib" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libaddr2line-08ae1606a951cabe.rlib" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgimli-682a81c4b2133b72.rlib" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_demangle-a73b3512c88de071.rlib" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_detect-7b5ec4c918d9f957.rlib" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libhashbrown-65c63cf3af0af657.rlib" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libminiz_oxide-4a53f0a2785abc6a.rlib" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libadler-868e2d515c28d027.rlib" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_alloc-394ad2d73aede76a.rlib" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-19c77e4dc3dcb87e.rlib" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcfg_if-fb44a42088c9369a.rlib" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-61a7402e61a5b0e0.rlib" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-63f8356c87a0d0e8.rlib" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/librustc_std_workspace_core-a506e577d917828c.rlib" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-fc1fb63210fdafad.rlib" "-Wl,--end-group" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-c21be34a5cae8449.rlib" "-Wl,-Bdynamic" "-lssl" "-lcrypto" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-znoexecstack" "-L" "/root/.rustup/toolchains/1.63-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libaziot_key_openssl_engine_shared.so" "-Wl,--gc-sections" "-shared" "-Wl,-zrelro,-znow" "-Wl,-O1" "-nodefaultlibs"
  = note: /tmp/ccJcIeVJ.s: Assembler messages:
          /tmp/ccJcIeVJ.s: Error: invalid attempt to declare external version name as default in symbol `bind_engine@@'
          lto-wrapper: fatal error: cc returned 1 exit status
          compilation terminated.
          /usr/bin/ld: error: lto-wrapper failed
          collect2: error: ld returned 1 exit status

error: could not compile `aziot-key-openssl-engine-shared` due to previous error

Caused by:
  process didn't exit successfully: `rustc --crate-name aziot_key_openssl_engine_shared --edition=2021 key/aziot-key-openssl-engine-shared/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --crate-type cdylib --emit=dep-info,link -C opt-level=3 -C panic=abort -C embed-bitcode=no -C metadata=961ba8918ce941b6 --out-dir /tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps -L dependency=/tmp/aziot-identity-service-1.4.0/target/release/deps --extern aziot_key_client=/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libaziot_key_client-e151afd6f6366504.rlib --extern aziot_key_common_http=/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libaziot_key_common_http-386aafcfd9410b71.rlib --extern aziot_key_openssl_engine=/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libaziot_key_openssl_engine-3fc2df14767b1cd3.rlib --extern http_common=/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libhttp_common-aa536d2159557dad.rlib --extern openssl_errors=/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libopenssl_errors-787e9d1b6661eeb1.rlib --extern openssl_sys=/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libopenssl_sys-ff039b1d2e627490.rlib --extern openssl_sys2=/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/libopenssl_sys2-d808bccdf70921d9.rlib --extern url=/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/deps/liburl-e7912308b310ec61.rlib -L native=/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/build/aziot-key-openssl-engine-shared-e763449c4ea9bf98/out -l 'static:+whole-archive=aziot_key_openssl_shared_engine_wrapper' -L native=/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/build/aziot-key-openssl-engine-950e6f910dc97734/out -L native=/tmp/aziot-identity-service-1.4.0/target/x86_64-unknown-linux-gnu/release/build/openssl-sys2-1ecd412231953aa1/out --cfg ossl300 --cfg ossl110` (exit status: 1)
warning: build failed, waiting for other jobs to finish...
make[3]: *** [Makefile:83: default] Error 101
make[3]: Leaving directory '/tmp/aziot-identity-service-1.4.0'
make[2]: *** [debian/rules:30: override_dh_auto_build] Error 2
make[2]: Leaving directory '/tmp/aziot-identity-service-1.4.0'
make[1]: *** [debian/rules:19: build] Error 2
make[1]: Leaving directory '/tmp/aziot-identity-service-1.4.0'
dpkg-buildpackage: error: debian/rules build subprocess returned exit status 2
make: *** [Makefile:312: deb] Error 2
```

Reply:
Odd... now that I've added the arm32/64 builds it seems it only happens for amd64 / arm64.  Arm32 is OK.  Any chance it has something to do with bitness?  I'm at a loss.

Reply:
@arsing remind me... the standalone engine is provided as a convenience for if someone wanted to use the C SDK and load the engine to interact with the keys service.  Neither the other daemons or IoT Edge depend on it, correct?  Given the changes in openssl 3 to move away from engines altogether I'm inclined to conditionally skip building and not include the engine in the packages for 22.04.

Reply:
No, all our services that do TLS (certd and identityd) also use the engine. They couldn't do TLS with keyd-backed keys otherwise. In fact the engine started out as something internal just for us, but I was convinced to make it available for users as well.

Reply:
OK then the workaround I see at the moment is to build the bits before dpkg-buildpackage and have the packaging build step be a no-op so it skips right to the install.  I haven't figured out why building as part of dpkg-buildpackage on 22.04 triggers the error.

Reply:
With that workaround everything appears to be working now.